akka-stream/src/main/scala/akka/stream/javadsl/Source.scala

```diff
@@ -797,7 +797,36 @@ final class Source[+Out, +Mat](delegate: scaladsl.Source[Out, Mat]) extends Grap
    * '''Cancels when''' downstream cancels
    */
   def mapConcat[T](f: function.Function[Out, _ <: java.lang.Iterable[T]]): javadsl.Source[T, Mat] =
-    new Source(delegate.mapConcat(elem ⇒ Util.immutableSeq(f.apply(elem))))
+    new Source(delegate.statefulMapConcat(() ⇒ elem ⇒ Util.immutableSeq(f.apply(elem))))
+
+
+  /**
+    * Transform each input element into an `Iterable` of output elements that is
+    * then flattened into the output stream. Transformation is stateful
+    * meaning it's taking into account previous messages (hysteresis).
+    *
+    * Make sure that the `Iterable` is immutable or at least not modified after
+    * being used as an output sequence. Otherwise the stream may fail with
+    * `ConcurrentModificationException` or other more subtle errors may occur.
+    *
+    * The returned `Iterable` MUST NOT contain `null` values,
+    * as they are illegal as stream elements - according to the Reactive Streams specification.
+    *
+    * '''Emits when''' the mapping function returns an element or there are still remaining elements
+    * from the previously calculated collection
+    *
+    * '''Backpressures when''' downstream backpressures or there are still remaining elements from the
+    * previously calculated collection
+    *
+    * '''Completes when''' upstream completes and all remaining elements has been emitted
+    *
+    * '''Cancels when''' downstream cancels
+    */
+  def statefulMapConcat[T](f: function.Creator[function.Function[Out, java.lang.Iterable[T]]]): javadsl.Source[T, Mat] =
+    new Source(delegate.statefulMapConcat{ () ⇒ {
+      val fun = f.create()
+      elem ⇒ Util.immutableSeq(fun(elem))
+    }})
```

Comment:
As you are not allowed to mutate values outside of the closure with Java lambdas, would it maybe make more sense with a specific type to pass in here?